Question: Anyone had experience of using kgo driver with Google's PubSubLite Kafka wire protocol API?

Thanks for the fantastic library @twmb . I wonder if there are folks out there using `kgo` to interact with PubSubLite (PSL) of GCP? Any experiences that could be shared would be appreciated!

We are testing out... and encountered some incompatibility, especially with consumer group and ListOffsets requests. In particular, we try to find a consumer group's committed offsets for a topic (https://pkg.go.dev/github.com/twmb/franz-go/pkg/kadm#Client.FetchOffsetsForTopics). PSL always returns -1. As a background, a PSL subscription is kind of a consumer group on a topic. And here is what GCP PSL folks told us the challenges on their end (they claim to conform to the Confluent Kafka protocol specs):
```
...
In particular, the ListOffsetsForTopics method does not just make a kafka OffsetFetchRequest with the consumer group and
offsets: instead it makes a list topic request, then makes an OffsetFetchRequest for [*all* topics for the consumer group]
(https://github.com/twmb/franz-go/blob/b13e4c4c6074dbab4699dc36f55bc09d1c1c2e2f/pkg/kadm/groups.go#L841), and
discards results that don't match the topics it listed. This makes no sense to me, as it would be easier to just make the single
[OffsetFetchRequest](https://kafka.apache.org/protocol.html#The_Messages_OffsetFetch) which takes a topic list.

We do not have the concept of a "multi topic consumer group" in Pub/Sub Lite, thus we have substituted the subscription
everywhere a topic is needed on the consumer group side APIs. This does mean however that we cannot service an
OffsetFetchRequest with no topics.
```
@twmb what do you think? If `kgo` can be made to be more compatible with PubSubLite, that will be more fantastic.

We also encountered a problem after v1.13.5 when a reader tries to join a consumer group, it had a problem with error:
```unable to join group session: ILLEGAL_GENERATION: Specified group generation id is not valid.```
And eventually it was successful. (only one consumer reader in the group)
On v1.13.5, it just hang without any error.

Thanks!

Thanks @twmb . A fundamental "limitation" with PSL as they say, is `subscription = a consumergroup on a single topic`, and one consumer group can have only one topic. All requests with consumer group is replaced with a subscription (on a topic).

I've probably mixed two different offsets in the discussions. Our usage patterns have two: one can use a consumer group without caring about which partitions messages come from; another requires reading from specific partitions as logs. We have a need to fetch/list offsets for partition's first and last offsets, as well as commit offsets.
1) For commit offsets, `FetchOffsetsForTopics()` was not working as expected for PSL as mentioned above. I tested with manual OffsetFetchRequest, and going through `client.Request()`, it does return the commit offsets. So this is a good alternative (workaround). Thanks for the suggestion.

2) For partition-based reading, at start of a job, we check the available first and last offsets, so we use ListOffsets to find out. I saw kadm has ListStartOffsets(ctx, topics...) and ListEndOffsets(ctx, topics). Even I specify one topic, PSL didn't seem to return the correct EndOffsets. And I also manually constructed ListOffsetsRequest with a topic, and sent with `client.Request()` PSL has an EOF error. I assume `client.Request()` won't change the provided request. Even with a specified topic, PSL seems to have a problem. (The PSL folks' response above seemed to indicate topics are removed and then filtered from returned results)
Here is the test code snippet:
```
	// Construct ListOffsetRequest
	topicPartitionFirst0 := kmsg.NewListOffsetsRequestTopicPartition()
	topicPartitionFirst0.Partition = 0
	topicPartitionFirst0.Timestamp = FirstOffset
	topicPartitionLast0 := kmsg.NewListOffsetsRequestTopicPartition()
	topicPartitionLast0.Partition = 0
	topicPartitionLast0.Timestamp = LastOffset

	topicPartitionFirst1 := kmsg.NewListOffsetsRequestTopicPartition()
	topicPartitionFirst1.Partition = 0
	topicPartitionFirst1.Timestamp = FirstOffset
	topicPartitionLast1 := kmsg.NewListOffsetsRequestTopicPartition()
	topicPartitionLast1.Partition = 0
	topicPartitionLast1.Timestamp = LastOffset

	topicOffsetReq := kmsg.NewListOffsetsRequestTopic()
	topicOffsetReq.Topic = topic
	topicOffsetReq.Partitions = []kmsg.ListOffsetsRequestTopicPartition{
		topicPartitionFirst0,
		topicPartitionLast0,
		topicPartitionFirst1,
		topicPartitionLast1,
	}

	listOffsetReq := kmsg.NewListOffsetsRequest()
	listOffsetReq.Topics = []kmsg.ListOffsetsRequestTopic{topicOffsetReq}

	resp, err := client.Request(context.Background(), &listOffsetReq) // return Response
	if err != nil {
		fmt.Println("list offset", "error", err, "partitions", 0, 1)
		return
	}
```
Debug dump:
```
dialing the broker
client created
[DEBUG] opening connection to broker; addr: us-west1-kafka-pubsub.googleapis.com:443, broker: seed_0
[DEBUG] connection opened to broker; addr: us-west1-kafka-pubsub.googleapis.com:443, broker: seed_0
[DEBUG] issuing api versions request; broker: seed_0, version: 3
[DEBUG] wrote ApiVersions v3; broker: seed_0, bytes_written: 31, write_wait: 25.482µs, time_to_write: 60.307µs, err: <nil>
[DEBUG] read ApiVersions v3; broker: seed_0, bytes_read: 121, read_wait: 57.926µs, time_to_read: 93.87306ms, err: <nil>
[DEBUG] issuing SASLHandshakeRequest; broker: seed_0
[DEBUG] wrote SASLHandshake v1; broker: seed_0, bytes_written: 24, write_wait: 23.095µs, time_to_write: 95.883µs, err: <nil>
[DEBUG] read SASLHandshake v1; broker: seed_0, bytes_read: 34, read_wait: 28.925µs, time_to_read: 28.827263ms, err: <nil>
[DEBUG] beginning sasl authentication; broker: seed_0, addr: us-west1-kafka-pubsub.googleapis.com:443, mechanism: PLAIN, authenticate: true
[DEBUG] issuing SASLAuthenticate; broker: seed_0, version: 0, step: 0
[DEBUG] wrote SASLAuthenticate v0; broker: seed_0, bytes_written: 3171, write_wait: 21.819µs, time_to_write: 100.298µs, err: <nil>
[DEBUG] read SASLAuthenticate v0; broker: seed_0, bytes_read: 16, read_wait: 25.655µs, time_to_read: 112.596419ms, err: <nil>
[DEBUG] connection initialized successfully; addr: us-west1-kafka-pubsub.googleapis.com:443, broker: seed_0
[DEBUG] wrote ApiVersions v0; broker: seed_0, bytes_written: 17, write_wait: 561.566766ms, time_to_write: 78.757µs, err: <nil>
[DEBUG] read ApiVersions v0; broker: seed_0, bytes_read: 104, read_wait: 52.064µs, time_to_read: 26.423359ms, err: <nil>
ping successful
[DEBUG] wrote Metadata v1; broker: seed_0, bytes_written: 97, write_wait: 24.805µs, time_to_write: 66.166µs, err: <nil>
[DEBUG] read Metadata v1; broker: seed_0, bytes_read: 311, read_wait: 53.498µs, time_to_read: 97.933226ms, err: <nil>
[DEBUG] sharded request; req: ListOffsets, destinations: [601576412]
[DEBUG] opening connection to broker; addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, broker: 601576412
[DEBUG] connection opened to broker; addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, broker: 601576412
[DEBUG] issuing api versions request; broker: 601576412, version: 3
[DEBUG] wrote ApiVersions v3; broker: 601576412, bytes_written: 31, write_wait: 8.396µs, time_to_write: 40.008µs, err: <nil>
[DEBUG] read ApiVersions v3; broker: 601576412, bytes_read: 121, read_wait: 29.666µs, time_to_read: 4.186362683s, err: <nil>
[DEBUG] issuing SASLHandshakeRequest; broker: 601576412
[DEBUG] wrote SASLHandshake v1; broker: 601576412, bytes_written: 24, write_wait: 12.314µs, time_to_write: 70.745µs, err: <nil>
[DEBUG] read SASLHandshake v1; broker: 601576412, bytes_read: 34, read_wait: 27.369µs, time_to_read: 1.612004859s, err: <nil>
[DEBUG] beginning sasl authentication; broker: 601576412, addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, mechanism: PLAIN, authenticate: true
[DEBUG] issuing SASLAuthenticate; broker: 601576412, version: 0, step: 0
[DEBUG] wrote SASLAuthenticate v0; broker: 601576412, bytes_written: 3171, write_wait: 12.596µs, time_to_write: 81.153µs, err: <nil>
[DEBUG] read SASLAuthenticate v0; broker: 601576412, bytes_read: 16, read_wait: 24.277µs, time_to_read: 4.0765286s, err: <nil>
[DEBUG] connection initialized successfully; addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, broker: 601576412
[DEBUG] wrote ListOffsets v4; broker: 601576412, bytes_written: 170, write_wait: 17.666475959s, time_to_write: 69.933µs, err: <nil>
[DEBUG] read ListOffsets v4; broker: 601576412, bytes_read: 0, read_wait: 72.942µs, time_to_read: 2.009925653s, err: EOF
[DEBUG] read from broker errored, killing connection; addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, broker: 601576412, successful_reads: 0, err: EOF
[DEBUG] sharded request failed, resharding and reissuing; req: ListOffsets, time_since_start: 21.776269151s, tries: 0, err: EOF
[DEBUG] opening connection to broker; addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, broker: 601576412
[DEBUG] connection opened to broker; addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, broker: 601576412
[DEBUG] issuing SASLHandshakeRequest; broker: 601576412
[DEBUG] wrote SASLHandshake v1; broker: 601576412, bytes_written: 24, write_wait: 22.15µs, time_to_write: 43.205µs, err: <nil>
[DEBUG] read SASLHandshake v1; broker: 601576412, bytes_read: 34, read_wait: 23.229µs, time_to_read: 61.762482ms, err: <nil>
[DEBUG] beginning sasl authentication; broker: 601576412, addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, mechanism: PLAIN, authenticate: true
[DEBUG] issuing SASLAuthenticate; broker: 601576412, version: 0, step: 0
[DEBUG] wrote SASLAuthenticate v0; broker: 601576412, bytes_written: 3171, write_wait: 19.298µs, time_to_write: 204.905µs, err: <nil>
[DEBUG] read SASLAuthenticate v0; broker: 601576412, bytes_read: 16, read_wait: 53.882µs, time_to_read: 93.00273ms, err: <nil>
[DEBUG] connection initialized successfully; addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, broker: 601576412
[DEBUG] wrote Metadata v1; broker: 601576412, bytes_written: 97, write_wait: 210.042753ms, time_to_write: 64.377µs, err: <nil>
[DEBUG] read Metadata v1; broker: 601576412, bytes_read: 311, read_wait: 32.152µs, time_to_read: 77.616231ms, err: <nil>
[DEBUG] sharded request; req: ListOffsets, destinations: [601576412]
[DEBUG] wrote ListOffsets v4; broker: 601576412, bytes_written: 170, write_wait: 41.742µs, time_to_write: 65.736µs, err: <nil>
[DEBUG] read ListOffsets v4; broker: 601576412, bytes_read: 0, read_wait: 40.728µs, time_to_read: 107.640338ms, err: EOF
[DEBUG] read from broker errored, killing connection; addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, broker: 601576412, successful_reads: 1, err: EOF
[DEBUG] sharded request failed, resharding and reissuing; req: ListOffsets, time_since_start: 26.172443652s, tries: 1, err: EOF
[DEBUG] wrote Metadata v1; broker: seed_0, bytes_written: 97, write_wait: 100.653µs, time_to_write: 73.381µs, err: <nil>
[DEBUG] read Metadata v1; broker: seed_0, bytes_read: 311, read_wait: 40.208µs, time_to_read: 74.868222ms, err: <nil>
[DEBUG] sharded request; req: ListOffsets, destinations: [601576412]
[DEBUG] opening connection to broker; addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, broker: 601576412
[DEBUG] connection opened to broker; addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, broker: 601576412
[DEBUG] issuing SASLHandshakeRequest; broker: 601576412
[DEBUG] wrote SASLHandshake v1; broker: 601576412, bytes_written: 24, write_wait: 8.022µs, time_to_write: 41.77µs, err: <nil>
[DEBUG] read SASLHandshake v1; broker: 601576412, bytes_read: 34, read_wait: 25.742µs, time_to_read: 125.252305ms, err: <nil>
[DEBUG] beginning sasl authentication; broker: 601576412, addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, mechanism: PLAIN, authenticate: true
[DEBUG] issuing SASLAuthenticate; broker: 601576412, version: 0, step: 0
[DEBUG] wrote SASLAuthenticate v0; broker: 601576412, bytes_written: 3171, write_wait: 12.089µs, time_to_write: 99.778µs, err: <nil>
[DEBUG] read SASLAuthenticate v0; broker: 601576412, bytes_read: 16, read_wait: 25.188µs, time_to_read: 113.316354ms, err: <nil>
[DEBUG] connection initialized successfully; addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, broker: 601576412
[DEBUG] wrote ListOffsets v4; broker: 601576412, bytes_written: 170, write_wait: 301.218695ms, time_to_write: 74.49µs, err: <nil>
[DEBUG] read ListOffsets v4; broker: 601576412, bytes_read: 0, read_wait: 39.3µs, time_to_read: 101.338719ms, err: EOF
[DEBUG] read from broker errored, killing connection; addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, broker: 601576412, successful_reads: 0, err: EOF
list offset error EOF partitions 0 1
```


3) For generation ID error during joining a group, later we found it could work sometimes without our notice. But I have the following debug dump.
options used for the client:
```
        maxPartBytes := int32(64e6) // 64MB
	tlsDialer := &tls.Dialer{NetDialer: &net.Dialer{Timeout: 10 * time.Second}}
	opts := []kgo.Opt{
		kgo.SeedBrokers(strings.Split(brokers, ",")...),
		kgo.ConsumeResetOffset(kgo.NewOffset().At(0)), // if offset is OutOfRange
		kgo.ConsumeTopics(topic),
		kgo.ConsumerGroup(consumerGroup),
		kgo.FetchMaxBytes(maxPartBytes),
		kgo.FetchMaxPartitionBytes(maxPartBytes),
		kgo.FetchMaxWait(time.Second),
		kgo.DisableAutoCommit(),
		kgo.RetryBackoffFn(func(n int) time.Duration {
			return 1 << (n % 6) * time.Second // up to 32 seconds
		}),
		kgo.WithLogger(kgo.BasicLogger(os.Stdout, kgo.LogLevelDebug, nil)),

		// SASL Options
		kgo.SASL(plain.Auth{
			User: user,
			Pass: password,
		}.AsMechanism()),

		// Configure TLS. Uses SystemCertPool for RootCAs by default.
		kgo.Dialer(tlsDialer.DialContext),
	}
```
Debug output: (the project number is redacted to 999999, nothing else touched)
```
dialing the broker
client created
[DEBUG] opening connection to broker; addr: us-west1-kafka-pubsub.googleapis.com:443, broker: seed_0
[INFO] immediate metadata update triggered; why: querying metadata for consumer initialization
[DEBUG] connection opened to broker; addr: us-west1-kafka-pubsub.googleapis.com:443, broker: seed_0
[DEBUG] issuing api versions request; broker: seed_0, version: 3
[DEBUG] wrote ApiVersions v3; broker: seed_0, bytes_written: 31, write_wait: 25.213µs, time_to_write: 64.328µs, err: <nil>
[DEBUG] read ApiVersions v3; broker: seed_0, bytes_read: 121, read_wait: 40.711µs, time_to_read: 90.902544ms, err: <nil>
[DEBUG] issuing SASLHandshakeRequest; broker: seed_0
[DEBUG] wrote SASLHandshake v1; broker: seed_0, bytes_written: 24, write_wait: 16.291µs, time_to_write: 72.758µs, err: <nil>
[DEBUG] read SASLHandshake v1; broker: seed_0, bytes_read: 34, read_wait: 23.459µs, time_to_read: 25.223858ms, err: <nil>
[DEBUG] beginning sasl authentication; broker: seed_0, addr: us-west1-kafka-pubsub.googleapis.com:443, mechanism: PLAIN, authenticate: true
[DEBUG] issuing SASLAuthenticate; broker: seed_0, version: 0, step: 0
[DEBUG] wrote SASLAuthenticate v0; broker: seed_0, bytes_written: 3171, write_wait: 21.218µs, time_to_write: 93.008µs, err: <nil>
[DEBUG] read SASLAuthenticate v0; broker: seed_0, bytes_read: 16, read_wait: 53.573µs, time_to_read: 96.156686ms, err: <nil>
[DEBUG] connection initialized successfully; addr: us-west1-kafka-pubsub.googleapis.com:443, broker: seed_0
[DEBUG] wrote ApiVersions v0; broker: seed_0, bytes_written: 17, write_wait: 263.599461ms, time_to_write: 75.754µs, err: <nil>
[DEBUG] wrote Metadata v1; broker: seed_0, bytes_written: 97, write_wait: 263.491697ms, time_to_write: 48.777µs, err: <nil>
[DEBUG] read ApiVersions v0; broker: seed_0, bytes_read: 104, read_wait: 173.979µs, time_to_read: 31.487455ms, err: <nil>
ping successful
[DEBUG] read Metadata v1; broker: seed_0, bytes_read: 311, read_wait: 31.744518ms, time_to_read: 67.250508ms, err: <nil>
[INFO] beginning to manage the group lifecycle; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub
[DEBUG] blocking commits from join&sync
[INFO] joining group; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub
[DEBUG] sharded request; req: FindCoordinator, destinations: [any]
[DEBUG] opening connection to broker; addr: 6543210.us-west1-kafka-pubsub.googleapis.com:443, broker: 6543210
[DEBUG] connection opened to broker; addr: 6543210.us-west1-kafka-pubsub.googleapis.com:443, broker: 6543210
[DEBUG] issuing api versions request; broker: 6543210, version: 3
[DEBUG] wrote ApiVersions v3; broker: 6543210, bytes_written: 31, write_wait: 15.299µs, time_to_write: 33.365µs, err: <nil>
[DEBUG] read ApiVersions v3; broker: 6543210, bytes_read: 121, read_wait: 35.054µs, time_to_read: 63.922143ms, err: <nil>
[DEBUG] issuing SASLHandshakeRequest; broker: 6543210
[DEBUG] wrote SASLHandshake v1; broker: 6543210, bytes_written: 24, write_wait: 20.108µs, time_to_write: 79.264µs, err: <nil>
[DEBUG] read SASLHandshake v1; broker: 6543210, bytes_read: 34, read_wait: 20.389µs, time_to_read: 29.620993ms, err: <nil>
[DEBUG] beginning sasl authentication; broker: 6543210, addr: 6543210.us-west1-kafka-pubsub.googleapis.com:443, mechanism: PLAIN, authenticate: true
[DEBUG] issuing SASLAuthenticate; broker: 6543210, version: 0, step: 0
[DEBUG] wrote SASLAuthenticate v0; broker: 6543210, bytes_written: 3171, write_wait: 12.339µs, time_to_write: 58.72µs, err: <nil>
[DEBUG] read SASLAuthenticate v0; broker: 6543210, bytes_read: 16, read_wait: 26.97µs, time_to_read: 81.334575ms, err: <nil>
[DEBUG] connection initialized successfully; addr: 6543210.us-west1-kafka-pubsub.googleapis.com:443, broker: 6543210
[DEBUG] sharded request failed, resharding and reissuing; req: FindCoordinator, time_since_start: 236.317357ms, tries: 0, err: broker is too old; the broker has already indicated it will not know how to handle the request
[DEBUG] sharded request; req: FindCoordinator, destinations: [any]
[DEBUG] opening connection to broker; addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, broker: 601576412
[DEBUG] connection opened to broker; addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, broker: 601576412
[DEBUG] issuing api versions request; broker: 601576412, version: 3
[DEBUG] wrote ApiVersions v3; broker: 601576412, bytes_written: 31, write_wait: 8.965µs, time_to_write: 39.214µs, err: <nil>
[DEBUG] read ApiVersions v3; broker: 601576412, bytes_read: 121, read_wait: 22.419µs, time_to_read: 78.209956ms, err: <nil>
[DEBUG] issuing SASLHandshakeRequest; broker: 601576412
[DEBUG] wrote SASLHandshake v1; broker: 601576412, bytes_written: 24, write_wait: 20.101µs, time_to_write: 51.711µs, err: <nil>
[DEBUG] read SASLHandshake v1; broker: 601576412, bytes_read: 34, read_wait: 26.435µs, time_to_read: 29.73184ms, err: <nil>
[DEBUG] beginning sasl authentication; broker: 601576412, addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, mechanism: PLAIN, authenticate: true
[DEBUG] issuing SASLAuthenticate; broker: 601576412, version: 0, step: 0
[DEBUG] wrote SASLAuthenticate v0; broker: 601576412, bytes_written: 3171, write_wait: 17.135µs, time_to_write: 83.713µs, err: <nil>
[DEBUG] read SASLAuthenticate v0; broker: 601576412, bytes_read: 16, read_wait: 32.505µs, time_to_read: 79.581843ms, err: <nil>
[DEBUG] connection initialized successfully; addr: 601576412.us-west1-kafka-pubsub.googleapis.com:443, broker: 601576412
[DEBUG] wrote FindCoordinator v1; broker: 601576412, bytes_written: 94, write_wait: 242.929313ms, time_to_write: 60.657µs, err: <nil>
[DEBUG] read FindCoordinator v1; broker: 601576412, bytes_read: 62, read_wait: 33.085µs, time_to_read: 22.345507ms, err: <nil>
[DEBUG] opening connection to broker; addr: 6543210.us-west1-kafka-pubsub.googleapis.com:443, broker: 6543210
[DEBUG] connection opened to broker; addr: 6543210.us-west1-kafka-pubsub.googleapis.com:443, broker: 6543210
[DEBUG] issuing SASLHandshakeRequest; broker: 6543210
[DEBUG] wrote SASLHandshake v1; broker: 6543210, bytes_written: 24, write_wait: 7.908µs, time_to_write: 44.185µs, err: <nil>
[DEBUG] read SASLHandshake v1; broker: 6543210, bytes_read: 34, read_wait: 24.625µs, time_to_read: 74.207018ms, err: <nil>
[DEBUG] beginning sasl authentication; broker: 6543210, addr: 6543210.us-west1-kafka-pubsub.googleapis.com:443, mechanism: PLAIN, authenticate: true
[DEBUG] issuing SASLAuthenticate; broker: 6543210, version: 0, step: 0
[DEBUG] wrote SASLAuthenticate v0; broker: 6543210, bytes_written: 3171, write_wait: 14.443µs, time_to_write: 101.703µs, err: <nil>
[DEBUG] read SASLAuthenticate v0; broker: 6543210, bytes_read: 16, read_wait: 25.296µs, time_to_read: 103.60413ms, err: <nil>
[DEBUG] connection initialized successfully; addr: 6543210.us-west1-kafka-pubsub.googleapis.com:443, broker: 6543210
[DEBUG] wrote JoinGroup v0; broker: 6543210, bytes_written: 241, write_wait: 234.083417ms, time_to_write: 52.845µs, err: <nil>
[DEBUG] read JoinGroup v0; broker: 6543210, bytes_read: 114, read_wait: 41.635µs, time_to_read: 32.307738ms, err: <nil>
[INFO] joined; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub, member_id: 7a749616-40e7-4480-9057-5c910dda1bb2, instance_id: <nil>, generation: -1, leader: false
[INFO] syncing; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub, protocol_type: consumer, protocol: cooperative-sticky
[DEBUG] wrote SyncGroup v0; broker: 6543210, bytes_written: 139, write_wait: 26.995µs, time_to_write: 64.549µs, err: <nil>
[DEBUG] read SyncGroup v0; broker: 6543210, bytes_read: 24, read_wait: 40.657µs, time_to_read: 29.077189ms, err: <nil>
[INFO] synced; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub, assigned: 
[DEBUG] unblocking commits from join&sync
[INFO] new group session begun; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub, added: , lost: 
[INFO] beginning heartbeat loop; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub
[DEBUG] entering OnPartitionsAssigned; with: map[]
[DEBUG] heartbeating; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub
[DEBUG] wrote Heartbeat v0; broker: 6543210, bytes_written: 135, write_wait: 72.538µs, time_to_write: 264.907µs, err: <nil>
[DEBUG] read Heartbeat v0; broker: 6543210, bytes_read: 10, read_wait: 88.161µs, time_to_read: 21.350199ms, err: <nil>
[DEBUG] heartbeat complete; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub, err: ILLEGAL_GENERATION: Specified group generation id is not valid.
[INFO] heartbeat errored; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub, err: ILLEGAL_GENERATION: Specified group generation id is not valid.
[DEBUG] entering OnPartitionsLost; with: map[]
[INFO] injecting fake fetch with an error; err: unable to join group session: ILLEGAL_GENERATION: Specified group generation id is not valid., why: notification of group management loop error
[INFO] assigning partitions; why: clearing assignment at end of group management session, how: unassigning everything, input: 
[ERROR] join and sync loop errored; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub, err: ILLEGAL_GENERATION: Specified group generation id is not valid., consecutive_errors: 1, backoff: 2s
Error from reader: unable to join group session: ILLEGAL_GENERATION: Specified group generation id is not valid.
[DEBUG] blocking commits from join&sync
[INFO] joining group; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub
[DEBUG] wrote JoinGroup v0; broker: 6543210, bytes_written: 273, write_wait: 31.085µs, time_to_write: 139.547µs, err: <nil>
[DEBUG] read JoinGroup v0; broker: 6543210, bytes_read: 114, read_wait: 48.222µs, time_to_read: 19.503683ms, err: <nil>
[INFO] joined; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub, member_id: 7a749616-40e7-4480-9057-5c910dda1bb2, instance_id: <nil>, generation: -1, leader: false
[INFO] syncing; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub, protocol_type: consumer, protocol: cooperative-sticky
[DEBUG] wrote SyncGroup v0; broker: 6543210, bytes_written: 139, write_wait: 16.731µs, time_to_write: 50.95µs, err: <nil>
[DEBUG] read SyncGroup v0; broker: 6543210, bytes_read: 24, read_wait: 30.239µs, time_to_read: 20.315476ms, err: <nil>
[INFO] synced; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub, assigned: 
[DEBUG] unblocking commits from join&sync
[INFO] new group session begun; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub, added: , lost: 
[INFO] beginning heartbeat loop; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub
[DEBUG] entering OnPartitionsAssigned; with: map[]
[DEBUG] heartbeating; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub
[DEBUG] wrote Heartbeat v0; broker: 6543210, bytes_written: 135, write_wait: 32.325µs, time_to_write: 168.002µs, err: <nil>
[DEBUG] read Heartbeat v0; broker: 6543210, bytes_read: 10, read_wait: 65.725µs, time_to_read: 24.083579ms, err: <nil>
[DEBUG] heartbeat complete; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub, err: ILLEGAL_GENERATION: Specified group generation id is not valid.
[INFO] heartbeat errored; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub, err: ILLEGAL_GENERATION: Specified group generation id is not valid.
[DEBUG] entering OnPartitionsLost; with: map[]
[INFO] injecting fake fetch with an error; err: unable to join group session: ILLEGAL_GENERATION: Specified group generation id is not valid., why: notification of group management loop error
[INFO] assigning partitions; why: clearing assignment at end of group management session, how: unassigning everything, input: 
[ERROR] join and sync loop errored; group: projects/999999/locations/us-west1-c/subscriptions/topic-test-gz-sub, err: ILLEGAL_GENERATION: Specified group generation id is not valid., consecutive_errors: 2, backoff: 4s
Error from reader: unable to join group session: ILLEGAL_GENERATION: Specified group generation id is not valid.
... (and it just repeats)
```

--- a/pkg/kadm/groups.go
+++ b/pkg/kadm/groups.go
@@ -830,6 +830,14 @@ func (cl *Client) FetchOffsets(ctx context.Context, group string) (OffsetRespons
 	return rs, nil
 }
 
+// FetchAllGroupTopics is a kadm "internal" topic name that can be used in
+// [FetchOffsetsForTopics]. By default, [FetchOffsetsForTopics] only returns
+// topics that are explicitly requested. Other topics that may be committed to
+// in the group are not returned. Using FetchAllRequestedTopics switches the
+// behavior to return the union of all committed topics and all requested
+// topics.
+const FetchAllGroupTopics = "|fetch-all-group-topics|"
+
 // FetchOffsetsForTopics is a helper function that returns the currently
 // committed offsets for the given group, as well as default -1 offsets for any
 // topic/partition that does not yet have a commit.
@@ -838,9 +846,31 @@ func (cl *Client) FetchOffsets(ctx context.Context, group string) (OffsetRespons
 // error. The returned offset responses are ready to be used or converted
 // directly to pure offsets with `Into`, and again into kgo offsets with
 // another `Into`.
+//
+// By default, this function returns offsets for only the requested topics. You
+// can use the special "topic" [FetchAllGroupTopics] to return all committed-to
+// topics in addition to all requested topics.
 func (cl *Client) FetchOffsetsForTopics(ctx context.Context, group string, topics ...string) (OffsetResponses, error) {
 	os := make(Offsets)
 
+	var all bool
+	keept := topics[:0]
+	for _, topic := range topics {
+		if topic == FetchAllGroupTopics {
+			all = true
+			continue
+		}
+		keept = append(keept, topic)
+	}
+	topics = keept
+
+	if !all && len(topics) == 0 {
+		return make(OffsetResponses), nil
+	}
+
+	// We have to request metadata to learn all partitions in all the
+	// topics. The default returned offset for all partitions is filled in
+	// to be -1.
 	if len(topics) > 0 {
 		listed, err := cl.ListTopics(ctx, topics...)
 		if err != nil {
@@ -865,11 +895,29 @@ func (cl *Client) FetchOffsetsForTopics(ctx context.Context, group string, topic
 	if err := resps.Error(); err != nil {
 		return nil, fmt.Errorf("offset fetches had a load error, first error: %w", err)
 	}
+
+	// For any topic (and any partition) we explicitly asked for, if the
+	// partition does not exist in the response, we fill the default -1
+	// from above.
 	os.Each(func(o Offset) {
 		if _, ok := resps.Lookup(o.Topic, o.Partition); !ok {
 			resps.Add(OffsetResponse{Offset: o})
 		}
 	})
+
+	// If we are not requesting all group offsets, then we strip any topic
+	// that was not explicitly requested.
+	if !all {
+		tset := make(map[string]struct{})
+		for _, t := range topics {
+			tset[t] = struct{}{}
+		}
+		for t := range resps {
+			if _, ok := tset[t]; !ok {
+				delete(resps, t)
+			}
+		}
+	}
 	return resps, nil
 }
 

